gitignore all key files in config/credentials

gregmolnar opened the pull request:
As mentioned in https://github.com/rails/rails/discussions/54773, it would be a good idea to gitignore all key files with a wildcard. That's how it is done in the dockerignore file too.

zzak commented:
Could you add a test please?

gregmolnar commented:
@zzak I added the test.

zzak commented on railties/lib/rails/generators/rails/app/templates/gitignore.tt:
```diff
@@ -35,3 +35,4 @@
 
 /public/assets
 <% end -%>
+/config/credentials/*.key
```
Rather than adding it here, why don't we generate it like we do with the master key?

https://github.com/rails/rails/blob/951ea8f75d9bf6d2999df2f0497347976dce9476/railties/lib/rails/generators/rails/app/app_generator.rb#L191-L205

https://github.com/rails/rails/blob/951ea8f75d9bf6d2999df2f0497347976dce9476/railties/lib/rails/generators/rails/master_key/master_key_generator.rb#L35-L41

gregmolnar replied:
I changed it to use that. It took slightly more code, but I made it in a separate commit so I can revert if the previous version is preferred.

zzak commented:
Wouldn't it make sense to use `Rails::Generators::CredentialsGenerator` instead? I thought there were separate commands for each key we are generating so it makes sense to update the gitignore there.

gregmolnar committed:
ignore all key files during credentials generation

gregmolnar commented:
The actual keygeneration happens in `Rails::Generators::EncryptionKeyFileGenerator` and you are right, that's  a much better place for this, because it will add the new gitignore entry even for old apps when they generate a new key, ie they introduce a new environment.

gregmolnar commented:
@zzak I changed this quite a bit. I moved the gitignore part into Rails::Generators::EncryptionKeyFileGenerator and calling it from there every time a key is generated. This made obsolete the calls to igoring the keys from the various parts in the codebase, so I removed those.

zzak commented on railties/lib/rails/generators/rails/encryption_key_file/encryption_key_file_generator.rb:
```diff
@@ -44,13 +46,18 @@ def ignore_key_file(key_path, ignore: key_ignore(key_path))
         def key_ignore(key_path)
-          [ "", "/#{key_path}", "" ].join("\n")
+          key_path = Pathname.new(key_path) unless key_path.is_a?(Pathname)
+          [ "", "# Ignore key files for decrypting credentials and more.", "/#{key_path.dirname.join("*.key")}", "" ].join("\n")
```
```suggestion
          <<~IGNORE.squish
            
            # Ignore key files for decrypting credentials and more.
            /#{key_path.dirname.join("*.key")}
            
          IGNORE
```

You might want to inspect the output to make sure indents, etc line up.

zzak commented:
Could you add a changelog?
